File: src/2021/day1/solver.ts
import { readFile } from "../../utils/utils";

const inputArr: number[] = readFile("2021", "day1")
  .toString()
  .split("\n")
  .map(x => Number(x));

const solver1 = (measurements: number[]): number => {
  let depthIncreasesNum = 0;

  for (let i = 0; i < measurements.length - 1; i++) {
    if (measurements[i] < measurements[i + 1]) depthIncreasesNum++;
  }

  return depthIncreasesNum;
};

const solver2 = (measurements: number[]): number => {
  let sumsLargerThanPrevious = 0;

  for (let i = 0; i < measurements.length - 3; i++) {
    const sum1 = measurements[i] + measurements[i + 1] + measurements[i + 2];
    const sum2 =
      measurements[i + 1] + measurements[i + 2] + measurements[i + 3];
    if (sum2 > sum1) sumsLargerThanPrevious++;
  }

  return sumsLargerThanPrevious;
};

console.log(solver1(inputArr));
console.log(solver2(inputArr));

File: cpp/2021/day1/solver.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
using namespace std;

int solver(vector<int> &measurements, int windowSize)
{
  int depthIncreasesNum = 0;

  for (int i = 0; i < measurements.size() - windowSize; ++i)
  {
    int sum1 = 0;
    int sum2 = 0;
    for (int j = 0; j < windowSize; ++j)
    {
      sum1 = sum1 + measurements[i + j];
      sum2 = sum2 + measurements[i + j + 1];
    }
    if (sum2 > sum1)
      depthIncreasesNum++;
  }

  return depthIncreasesNum;
}

int main()
{
  vector<int> measurements;
  int number;

  ifstream input_file("input.txt");

  while (input_file >> number)
  {
    measurements.push_back(number);
  }

  // for (int i : measurements)
  // {
  //   cout << i << ",";
  // }
  // cout << endl;

  cout << solver(measurements, 1) << endl;
  cout << solver(measurements, 3) << endl;

  input_file.close();
}